servicetalk-grpc-protobuf/src/main/java/io/servicetalk/grpc/protobuf/ProtoBufSerializationProvider.java

```diff
@@ -146,7 +150,22 @@ private static boolean isCompressed(Buffer buffer) throws SerializationException
 
                     final T t;
                     try {
-                        t = parser.parseFrom(toDeserialize.toNioBuffer(toDeserialize.readerIndex(), lengthOfData));
+                        final CodedInputStream in;
+                        if (toDeserialize.nioBufferCount() == 1) {
+                            in = CodedInputStream.newInstance(toDeserialize.toNioBuffer(toDeserialize.readerIndex(),
+                                    lengthOfData));
+                        } else {
+                            // Aggregated payload body may consist of multiple Buffers. In this case,
+                            // CompositeBuffer.toNioBuffer(idx, length) may return a single ByteBuffer (when requested
+                            // length < components[0].length) or create a new ByteBuffer and copy multiple components
+                            // into it. Later, proto parser will copy data from this temporary ByteBuffer again.
+                            // To avoid unnecessary copying, we use newCodedInputStream(buffers, lengthOfData).
+                            final ByteBuffer[] buffers = toDeserialize.toNioBuffers(toDeserialize.readerIndex(),
+                                    lengthOfData);
+                            in = buffers.length == 1 ? CodedInputStream.newInstance(buffers[0]) :
```

Comment:
Is this check necessary since we will only be in this `else` block if `toDeserialize.nioBufferCount() > 1`?

Reply:
Yes, this is described in the comment above. `toDeserialize.nioBufferCount()` does not account for a requested length. We may be in a situation when the `CompositeBuffer` has 2 or more buffers internally (and those buffers may contain multiple proto messages), but we request only a few bytes (single proto message) that could be read from the first internal buffer. In this case, `toDeserialize.toNioBuffers(..., length)` may return an array that has only the first internal `ByteBuffer`, not all the internal buffers.